cairo/ethereum/cancun/vm/instructions/system.cairo

```diff
@@ -1,58 +1,253 @@
+from starkware.cairo.common.uint256 import uint256_lt
+from starkware.cairo.common.alloc import alloc
+
 from ethereum_types.others import (
     ListTupleU256U256,
     ListTupleU256U256Struct,
     TupleU256U256,
     TupleU256U256Struct,
 )
-from ethereum.cancun.state import get_account, account_has_code_or_nonce, increment_nonce
-from ethereum_types.bytes import Bytes, BytesStruct, Bytes0
-from ethereum.cancun.vm.stack import push
-from ethereum.cancun.transactions import To, ToStruct
-from ethereum.utils.numeric import is_zero, U256_from_be_bytes20
-from src.utils.dict import hashdict_write, dict_copy
-// from ethereum.cancun.vm.interpreter import process_create_message
-from starkware.cairo.common.math_cmp import is_le
-from starkware.cairo.common.dict_access import DictAccess
-from starkware.cairo.common.registers import get_fp_and_pc
-from starkware.cairo.common.uint256 import uint256_lt
-from starkware.cairo.common.alloc import alloc
 
-const STACK_DEPTH_LIMIT = 1024;
-const MAX_CODE_SIZE = 0x6000;
+func generic_call{
+    process_message_label: felt*,
+    range_check_ptr,
+    bitwise_ptr: BitwiseBuiltin*,
+    keccak_ptr: KeccakBuiltin*,
+    poseidon_ptr: PoseidonBuiltin*,
+    evm: Evm,
+}(
+    gas: Uint,
+    value: U256,
+    caller: Address,
+    to: Address,
+    code_address: Address,
+    should_transfer_value: bool,
+    is_staticcall: bool,
+    memory_input_start_position: U256,
+    memory_input_size: U256,
+    memory_output_start: U256,
+    memory_output_size: U256,
+) -> EthereumException* {
+    alloc_locals;
+    let fp_and_pc = get_fp_and_pc();
+    local __fp__: felt* = fp_and_pc.fp_val;
+
+    tempvar return_data = Bytes(new BytesStruct(cast(0, felt*), 0));
+    EvmImpl.set_return_data(return_data);
+
+    let depth_too_deep = is_le(STACK_DEPTH_LIMIT, evm.value.message.value.depth.value);
+    if (depth_too_deep != 0) {
+        let gas = Uint(gas.value + evm.value.gas_left.value);
+        EvmImpl.set_gas_left(gas);
+        let stack = evm.value.stack;
+        tempvar zero = U256(new U256Struct(0, 0));
+        let err = push{stack=stack}(zero);
+        if (cast(err, felt) != 0) {
+            EvmImpl.set_stack(stack);
+            return err;
+        }
+        let ok = cast(0, EthereumException*);
+        return ok;
+    }
+
+    let memory = evm.value.memory;
+    with memory {
+        let calldata = memory_read_bytes(memory_input_start_position, memory_input_size);
+    }
+    EvmImpl.set_memory(memory);
+
+    let env = evm.value.env;
+    let state = env.value.state;
+    let account = get_account{state=state}(code_address);
+
+    EnvImpl.set_state{env=env}(state);
+    let env = env;
+    EvmImpl.set_env(env);
+    let evm = evm;
+
+    let code = account.value.code;
+
+    if (is_staticcall.value != 0) {
+        tempvar is_static = bool(1);
+    } else {
+        tempvar is_static = evm.value.message.value.is_static;
+    }
+    let is_static = is_static;
```

Comment:
```suggestion
    let is_static = bool(is_staticcall.value + evm.value.message.value.is_static)
```

Reply:
I don't think we should have is_static = bool(2) no?

Reply:
we only do != 0 checks, so it's not important (and in Cairo it's a lot simpler to do `a OR b` then doing nested if-elses conditions)